File: src/main/java/io/nshusa/rsam/binary/sprite/Sprite.java
package io.nshusa.rsam.binary.sprite;

import io.nshusa.rsam.binary.Archive;
import io.nshusa.rsam.graphics.render.Raster;
import io.nshusa.rsam.util.ByteBufferUtils;
import io.nshusa.rsam.util.HashUtils;

import java.awt.image.BufferedImage;
import java.awt.image.DataBufferInt;
import java.io.IOException;
import java.nio.ByteBuffer;

public final class Sprite {

    private int width;
    private int height;
    private int offsetX;
    private int offsetY;
    private int resizeWidth;
    private int resizeHeight;
    private int[] pixels;
    private int format;

    public Sprite() {

    }

    public Sprite(int width, int height) {
        this.pixels = new int[width * height];
        this.width = this.resizeWidth = width;
        this.height = this.resizeHeight = height;
    }

    public Sprite(int resizeWidth, int resizeHeight, int horizontalOffset, int verticalOffset, int width, int height, int format,
                  int[] pixels) {
        this.resizeWidth = resizeWidth;
        this.resizeHeight = resizeHeight;
        this.offsetX = horizontalOffset;
        this.offsetY = verticalOffset;
        this.width = width;
        this.height = height;
        this.format = format;
        this.pixels = pixels;
    }

    public static Sprite decode(Archive archive, int hash, int id) throws IOException {
        ByteBuffer dataBuf = archive.readFile(hash);
        ByteBuffer metaBuf = archive.readFile("index.dat");

        Sprite sprite = new Sprite();

        // position of the current image archive within the archive
        metaBuf.position(dataBuf.getShort() & 0xFFFF);

        // the maximum width the images in this archive can scale to
        sprite.setResizeWidth(metaBuf.getShort() & 0xFFFF);

        // the maximum height the images in this archive can scale to
        sprite.setResizeHeight(metaBuf.getShort() & 0xFFFF);

        // the number of colors that are used in this image archive (limit is 256 if one of the rgb values is 0 else its 255)
        int colours = metaBuf.get() & 0xFF;

        // the array of colors that can only be used in this archive
        int[] palette = new int[colours];

        for (int index = 0; index < colours - 1; index++) {
            int colour = ByteBufferUtils.readU24Int(metaBuf);
            // + 1 because index = 0 is for transparency, = 1 is a flag for opacity. (BufferedImage#OPAQUE)
            palette[index + 1] = colour == 0 ? 1 : colour;
        }

        for (int i = 0; i < id; i++) {
            // skip the current offsetX and offsetY
            metaBuf.position(metaBuf.position() + 2);

            // skip the current array of pixels
            dataBuf.position(dataBuf.position() + ((metaBuf.getShort() & 0xFFFF) * (metaBuf.getShort() & 0xFFFF)));

            // skip the current format
            metaBuf.position(metaBuf.position() + 1);
        }

        // offsets are used to reposition the sprite on an interface.
        sprite.setOffsetX(metaBuf.get() & 0xFF);
        sprite.setOffsetY(metaBuf.get() & 0xFF);

        // actual width of this sprite
        sprite.setWidth(metaBuf.getShort() & 0xFFFF);

        // actual height of this sprite
        sprite.setHeight(metaBuf.getShort() & 0xFFFF);

        // there are 2 ways the pixels can be written (0 or 1, 0 means the position is read horizontally, 1 means vertically)
        sprite.setFormat(metaBuf.get() & 0xFF);

        if (sprite.getFormat() != 0 && sprite.getFormat() != 1) {
            throw new IOException(String.format("Detected end of archive=%d id=%d or wrong format=%d", hash, id, sprite.getFormat()));
        }

        if (sprite.getWidth() > 765 || sprite.getHeight() > 765 || sprite.getWidth() <= 0 || sprite.getHeight() <= 0) {
            throw new IOException(String.format("Detected end of archive=%d id=%d", hash, id));
        }

        int[] raster = new int[sprite.getWidth() * sprite.getHeight()];

        if (sprite.getFormat() == 0) { // read horizontally
            for (int index = 0; index < raster.length; index++) {
                raster[index] = palette[dataBuf.get() & 0xFF];
            }
        } else if (sprite.getFormat() == 1) { // read vertically
            for (int x = 0; x < sprite.getWidth(); x++) {
                for (int y = 0; y < sprite.getHeight(); y++) {
                    raster[x + y * sprite.getWidth()] = palette[dataBuf.get() & 0xFF];
                }
            }
        }
        sprite.setPixels(raster);
        return sprite;
    }

    public static Sprite decode(Archive archive, String name, int id) throws IOException {
        return decode(archive, HashUtils.nameToHash(name.contains(".dat") ? name : name + ".dat"), id);
    }

    public void drawSprite(int x, int y) {
        x += offsetX;
        y += offsetY;
        int rasterClip = x + y * Raster.width;
        int imageClip = 0;
        int height = this.height;
        int width = this.width;
        int rasterOffset = Raster.width - width;
        int imageOffset = 0;

        if (y < Raster.getClipBottom()) {
            int dy = Raster.getClipBottom() - y;
            height -= dy;
            y = Raster.getClipBottom();
            imageClip += dy * width;
            rasterClip += dy * Raster.width;
        }

        if (y + height > Raster.getClipTop()) {
            height -= y + height - Raster.getClipTop();
        }

        if (x < Raster.getClipLeft()) {
            int dx = Raster.getClipLeft() - x;
            width -= dx;
            x = Raster.getClipLeft();
            imageClip += dx;
            rasterClip += dx;
            imageOffset += dx;
            rasterOffset += dx;
        }

        if (x + width > Raster.getClipRight()) {
            int dx = x + width - Raster.getClipRight();
            width -= dx;
            imageOffset += dx;
            rasterOffset += dx;
        }

        if (width > 0 && height > 0) {
            draw(Raster.raster, pixels, 0, imageClip, rasterClip, width, height, rasterOffset, imageOffset);
        }
    }

    private void draw(int raster[], int[] image, int colour, int sourceIndex, int destIndex, int width, int height, int destStep,
                      int sourceStep) {
        int minX = -(width >> 2);
        width = -(width & 3);

        for (int y = -height; y < 0; y++) {
            for (int x = minX; x < 0; x++) {
                colour = image[sourceIndex++];
                if (colour != 0) {
                    raster[destIndex++] = colour;
                } else {
                    destIndex++;
                }
                colour = image[sourceIndex++];

                if (colour != 0) {
                    raster[destIndex++] = colour;
                } else {
                    destIndex++;
                }
                colour = image[sourceIndex++];

                if (colour != 0) {
                    raster[destIndex++] = colour;
                } else {
                    destIndex++;
                }
                colour = image[sourceIndex++];

                if (colour != 0) {
                    raster[destIndex++] = colour;
                } else {
                    destIndex++;
                }
            }

            for (int k2 = width; k2 < 0; k2++) {
                colour = image[sourceIndex++];
                if (colour != 0) {
                    raster[destIndex++] = colour;
                } else {
                    destIndex++;
                }
            }

            destIndex += destStep;
            sourceIndex += sourceStep;
        }
    }

    public BufferedImage toBufferedImage() {

        BufferedImage image = new BufferedImage(this.width, this.height, BufferedImage.TYPE_INT_RGB);

        final int[] pixels = ((DataBufferInt) image.getRaster().getDataBuffer()).getData();

        System.arraycopy(this.pixels, 0, pixels, 0, this.pixels.length);

        return image;
    }

    public int getWidth() {
        return width;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    public int getHeight() {
        return height;
    }

    public void setHeight(int height) {
        this.height = height;
    }

    public int getOffsetX() {
        return offsetX;
    }

    public void setOffsetX(int offsetX) {
        this.offsetX = offsetX;
    }

    public int getOffsetY() {
        return offsetY;
    }

    public void setOffsetY(int offsetY) {
        this.offsetY = offsetY;
    }

    public int getResizeWidth() {
        return resizeWidth;
    }

    public void setResizeWidth(int resizeWidth) {
        this.resizeWidth = resizeWidth;
    }

    public int getResizeHeight() {
        return resizeHeight;
    }

    public void setResizeHeight(int resizeHeight) {
        this.resizeHeight = resizeHeight;
    }

    public int[] getPixels() {
        return pixels;
    }

    public void setPixels(int[] pixels) {
        this.pixels = pixels;
    }

    public int getFormat() {
        return format;
    }

    public void setFormat(int format) {
        this.format = format;
    }

}

File: src/main/kotlin/io/nshusa/rsam/binary/sprite/Sprite.kt
package io.nshusa.rsam.binary.sprite

import io.nshusa.rsam.binary.Archive
import io.nshusa.rsam.graphics.render.Raster
import io.nshusa.rsam.util.ByteBufferUtils
import io.nshusa.rsam.util.HashUtils
import java.awt.Color
import java.awt.Image
import java.awt.Toolkit
import java.awt.image.BufferedImage
import java.awt.image.FilteredImageSource
import java.awt.image.RGBImageFilter
import java.io.IOException
import java.nio.ByteBuffer





class Sprite {

    var id: Int = 0
        private set
    var archive: Int = 0
        private set
    var width: Int = 0
    var height: Int = 0
    var offsetX: Int = 0
    var offsetY: Int = 0
    var resizeWidth: Int = 0
    var resizeHeight: Int = 0
    var pixels: IntArray? = null
    var format: Int = 0

    constructor()

    constructor(width: Int, height: Int) {
        this.pixels = IntArray(width * height)
        this.resizeWidth = width
        this.width = this.resizeWidth
        this.resizeHeight = height
        this.height = this.resizeHeight
    }

    constructor(resizeWidth: Int, resizeHeight: Int, horizontalOffset: Int, verticalOffset: Int, width: Int, height: Int, format: Int, pixels: IntArray) {
        this.resizeWidth = resizeWidth
        this.resizeHeight = resizeHeight
        this.offsetX = horizontalOffset
        this.offsetY = verticalOffset
        this.width = width
        this.height = height
        this.format = format
        this.pixels = pixels
    }

    fun drawSprite(x: Int, y: Int) {
        var x = x
        var y = y
        x += offsetX
        y += offsetY
        var rasterClip = x + y * Raster.width
        var imageClip = 0
        var height = this.height
        var width = this.width
        var rasterOffset = Raster.width - width
        var imageOffset = 0

        if (y < Raster.clipBottom) {
            val dy = Raster.clipBottom - y
            height -= dy
            y = Raster.clipBottom
            imageClip += dy * width
            rasterClip += dy * Raster.width
        }

        if (y + height > Raster.clipTop) {
            height -= y + height - Raster.clipTop
        }

        if (x < Raster.clipLeft) {
            val dx = Raster.clipLeft - x
            width -= dx
            x = Raster.clipLeft
            imageClip += dx
            rasterClip += dx
            imageOffset += dx
            rasterOffset += dx
        }

        if (x + width > Raster.clipRight) {
            val dx = x + width - Raster.clipRight
            width -= dx
            imageOffset += dx
            rasterOffset += dx
        }

        if (width > 0 && height > 0) {
            draw(Raster.raster, pixels, 0, imageClip, rasterClip, width, height, rasterOffset, imageOffset)
        }
    }

    private fun draw(raster: IntArray, image: IntArray?, colour: Int, sourceIndex: Int, destIndex: Int, width: Int, height: Int, destStep: Int,
                     sourceStep: Int) {
        var colour = colour
        var sourceIndex = sourceIndex
        var destIndex = destIndex
        var width = width
        val minX = -(width shr 2)
        width = -(width and 3)

        for (y in -height..-1) {
            for (x in minX..-1) {
                colour = image!![sourceIndex++]
                if (colour != 0) {
                    raster[destIndex++] = colour
                } else {
                    destIndex++
                }
                colour = image[sourceIndex++]

                if (colour != 0) {
                    raster[destIndex++] = colour
                } else {
                    destIndex++
                }
                colour = image[sourceIndex++]

                if (colour != 0) {
                    raster[destIndex++] = colour
                } else {
                    destIndex++
                }
                colour = image[sourceIndex++]

                if (colour != 0) {
                    raster[destIndex++] = colour
                } else {
                    destIndex++
                }
            }

            for (k2 in width..-1) {
                colour = image!![sourceIndex++]
                if (colour != 0) {
                    raster[destIndex++] = colour
                } else {
                    destIndex++
                }
            }

            destIndex += destStep
            sourceIndex += sourceStep
        }
    }

    private fun makeColorTransparent(im: BufferedImage, color: Color): Image {
        val filter = object : RGBImageFilter() {
            var markerRGB = color.rgb or -0x1000000
            override fun filterRGB(x: Int, y: Int, rgb: Int): Int {
                return if (rgb or -0x1000000 == markerRGB)
                    0x00FFFFFF and rgb
                else
                    rgb
            }
        }
        val ip = FilteredImageSource(im.source, filter)
        return Toolkit.getDefaultToolkit().createImage(ip)
    }

    private fun imageToBufferedImage(image: Image): BufferedImage {
        val bufferedImage = BufferedImage(image.getWidth(null), image.getHeight(null), BufferedImage.TYPE_INT_ARGB)
        val g2 = bufferedImage.createGraphics()
        g2.drawImage(image, 0, 0, null)
        g2.dispose()
        return bufferedImage
    }

    fun toBufferedImage(): BufferedImage {
        val bi = BufferedImage(width, height, BufferedImage.TYPE_INT_RGB)
        bi.setRGB(0, 0, width, height, pixels!!, 0, width)
        val img = makeColorTransparent(bi, Color(0, 0, 0))

        return imageToBufferedImage(img)
    }

    companion object {
        @Throws(IOException::class)
        fun decode(archive: Archive, metaBuf: ByteBuffer = archive.readFile("index.dat"), hash: Int, id: Int): Sprite {
            val dataBuf = archive.readFile(hash)

            val sprite = Sprite()
            sprite.id = id
            sprite.archive = hash

            // position of the current image archive within the archive
            metaBuf.position(dataBuf.short.toInt() and 0xFFFF)

            // the maximum width the images in this archive can scale to
            sprite.resizeWidth = metaBuf.short.toInt() and 0xFFFF

            // the maximum height the images in this archive can scale to
            sprite.resizeHeight = metaBuf.short.toInt() and 0xFFFF

            // the number of colors that are used in this image archive (limit is 256 if one of the rgb values is 0 else its 255)
            val colours = metaBuf.get().toInt() and 0xFF

            // the array of colors that can only be used in this archive
            val palette = IntArray(colours)

            for (index in 0 until colours - 1) {
                val colour = ByteBufferUtils.readU24Int(metaBuf)
                // + 1 because index = 0 is for transparency, = 1 is a flag for opacity. (BufferedImage#OPAQUE)
                palette[index + 1] = if (colour == 0) 1 else colour
            }

            for (i in 0 until id) {
                // skip the current offsetX and offsetY
                metaBuf.position(metaBuf.position() + 2)

                // skip the current array of pixels
                dataBuf.position(dataBuf.position() + (metaBuf.short.toInt() and 0xFFFF) * (metaBuf.short.toInt() and 0xFFFF))

                // skip the current format
                metaBuf.position(metaBuf.position() + 1)
            }

            // offsets are used to reposition the sprite on an interface.
            sprite.offsetX = metaBuf.get().toInt() and 0xFF
            sprite.offsetY = metaBuf.get().toInt() and 0xFF

            // actual width of this sprite
            sprite.width = metaBuf.short.toInt() and 0xFFFF

            // actual height of this sprite
            sprite.height = metaBuf.short.toInt() and 0xFFFF

            // there are 2 ways the pixels can be written (0 or 1, 0 means the position is read horizontally, 1 means vertically)
            sprite.format = metaBuf.get().toInt() and 0xFF

            if (sprite.format != 0 && sprite.format != 1) {
                throw IOException(String.format("Detected end of archive=%d id=%d or wrong format=%d", hash, id, sprite.format))
            }

            if (sprite.width > 765 || sprite.height > 765 || sprite.width <= 0 || sprite.height <= 0) {
                throw IOException(String.format("Detected end of archive=%d id=%d", hash, id))
            }

            val raster = IntArray(sprite.width * sprite.height)

            if (sprite.format == 0) { // read horizontally
                for (index in raster.indices) {
                    raster[index] = palette[dataBuf.get().toInt() and 0xFF]
                }
            } else if (sprite.format == 1) { // read vertically
                for (x in 0 until sprite.width) {
                    for (y in 0 until sprite.height) {
                        raster[x + y * sprite.width] = palette[dataBuf.get().toInt() and 0xFF]
                    }
                }
            }
            sprite.pixels = raster
            return sprite
        }


        @Throws(IOException::class)
        fun decode(archive: Archive, name: String, id: Int): Sprite {
            return decode(archive, hash = HashUtils.nameToHash(if (name.contains(".dat")) name else "$name.dat"), id = id)
        }
    }

}